File: prep/implementation/practice_iterations/merge_intervals_2/insert_interval.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

struct Interval {
  int start;
  int end;

  Interval(int start, int end) : start(start), end(end) {}
};

// O(n) / O(1)
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n;
  cin >> tt;

  while (tt--) {
    cin >> n;

    vector<Interval> a;
    int s, e;
    for (int i = 0; i < n; i++) {
      cin >> s >> e;
      a.push_back(Interval(s, e));
    }

    cin >> s >> e;
    Interval new_interval(s, e);

    int i = 0;
    vector<Interval> res;
    while (i < n && a[i].end < new_interval.start) {
      res.push_back(a[i++]);
    }

    while (i < n && a[i].start <= new_interval.end) {
      new_interval.start = min(new_interval.start, a[i].start);
      new_interval.end = max(new_interval.end, a[i].end);
      i++;
    }

    res.push_back(new_interval);

    while (i < n) {
      res.push_back(a[i++]);
    }

    for (auto inter : res) {
      cout << inter.start << " " << inter.end << "\n";
    }
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/iteration_archives/merge_intervals_4/InsertInterval.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(n) / O(1)
public class InsertInterval {
  static class Interval {
    int start;
    int end;

    Interval(int start, int end) {
      this.start = start;
      this.end = end;
    }
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        List<Interval> a = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
          int start = in.nextInt();
          int end = in.nextInt();
          a.add(new Interval(start, end));
        }

        int start = in.nextInt();
        int end = in.nextInt();
        Interval newInterval = new Interval(start, end);

        List<Interval> res = new ArrayList<>();

        int i = 0;
        while (i < n && a.get(i).end < newInterval.start) {
          res.add(a.get(i++));
        }

        while (i < n && a.get(i).start <= newInterval.end) {
          newInterval.start = Math.min(newInterval.start, a.get(i).start);
          newInterval.end = Math.max(newInterval.end, a.get(i).end);
          i++;
        }

        res.add(newInterval);

        while (i < n) {
          res.add(a.get(i++));
        }

        for (Interval inter : res) {
          System.out.printf("%d %d\n", inter.start, inter.end);
        }
      }
    }
  }
}